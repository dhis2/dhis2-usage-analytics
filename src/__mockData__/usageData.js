export const dataStatistics = [
    {
        year: 2016,
        week: 22,
        mapViews: 4457,
        chartViews: 5047,
        pivotTableViews: 4281,
        eventReportViews: 4649,
        eventChartViews: 4716,
        dashboardViews: 6460,
        dataSetReportViews: 4414,
        totalViews: 23669,
        averageViews: 104,
        averageMapViews: 19,
        averageChartViews: 22,
        averagePivotTableViews: 18,
        averageEventReportViews: 20,
        averageEventChartViews: 20,
        averageDashboardViews: 28,
        savedMaps: 1658,
        savedCharts: 1723,
        savedPivotTables: 1642,
        savedEventReports: 1748,
        savedEventCharts: 1613,
        savedDashboards: 1873,
        savedIndicators: 1826,
        savedDataValues: 11751,
        activeUsers: 227,
        users: 262,
    },
    {
        year: 2016,
        week: 23,
        mapViews: 4093,
        chartViews: 5179,
        pivotTableViews: 4275,
        eventReportViews: 4856,
        eventChartViews: 4782,
        dashboardViews: 6193,
        dataSetReportViews: 4242,
        totalViews: 23840,
        averageViews: 101,
        averageMapViews: 17,
        averageChartViews: 22,
        averagePivotTableViews: 18,
        averageEventReportViews: 20,
        averageEventChartViews: 20,
        averageDashboardViews: 26,
        savedMaps: 1593,
        savedCharts: 1668,
        savedPivotTables: 1643,
        savedEventReports: 1745,
        savedEventCharts: 1672,
        savedDashboards: 1853,
        savedIndicators: 1905,
        savedDataValues: 11520,
        activeUsers: 234,
        users: 262,
    },
]

export const topFavorites = [
    {
        position: 1,
        name: 'ANC: ANC 3 coverage this year (gauge)',
        views: 1,
        id: 'dw62KezChat',
        created: '2014-11-10T00:00:00.000',
    },
    {
        position: 2,
        name: 'Immunization: IPT 1 and 2 Coverage',
        views: 1,
        id: 'nwqNUFG9Vky',
        created: '2012-07-31T00:00:00.000',
    },
]
