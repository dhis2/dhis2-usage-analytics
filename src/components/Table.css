.uaa-data-table {
    width: 100%;
}
.uaa-data-table thead {
    border-bottom: 2px solid #e0e0e0;
}

.uaa-data-table tbody tr {
    border-bottom: 1px solid #e0e0e0;
}

.uaa-data-table th,
.uaa-data-table td {
    font-size: 14px;
    padding: 16px 16px;
}
.uaa-data-table th {
    text-align: left;
    font-weight: bold;
}
