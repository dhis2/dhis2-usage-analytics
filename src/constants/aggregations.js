import i18n from '@dhis2/d2-i18n'

export const SUM = 'SUM'
export const AVERAGE = 'AVERAGE'

const AGGREGATIONS = {
    [SUM]: { value: SUM, label: i18n.t('Sum') },
    [AVERAGE]: { value: AVERAGE, label: i18n.t('Average') },
}
export default AGGREGATIONS
