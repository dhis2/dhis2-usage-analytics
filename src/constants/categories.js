import i18n from '@dhis2/d2-i18n'

export const FAVORITE_VIEWS = 'FAVORITE_VIEWS'
export const FAVORITES_SAVED = 'FAVORITES_SAVED'
export const USERS = 'USERS'
export const TOP_FAVORITES = 'TOP_FAVORITES'
export const DATA_VALUES = 'DATA_VALUES'

const CATEGORIES = [
    { value: FAVORITE_VIEWS, label: i18n.t('Favorite views') },
    { value: FAVORITES_SAVED, label: i18n.t('Favorites saved') },
    { value: USERS, label: i18n.t('Users') },
    { value: TOP_FAVORITES, label: i18n.t('Top favorites') },
    { value: DATA_VALUES, label: i18n.t('Data values') },
]
export default CATEGORIES
