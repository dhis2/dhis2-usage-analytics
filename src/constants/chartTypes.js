import i18n from '@dhis2/d2-i18n'

export const ALL = 'ALL'
export const TOTAL = 'TOTAL'

const chartTypes = [
    { value: ALL, label: i18n.t('All') },
    { value: TOTAL, label: i18n.t('Total') },
]

export default chartTypes
