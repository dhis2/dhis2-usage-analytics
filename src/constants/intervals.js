import i18n from '@dhis2/d2-i18n'

export const DAY = 'DAY'
export const WEEK = 'WEEK'
export const MONTH = 'MONTH'
export const YEAR = 'YEAR'

const intervals = [
    { value: DAY, label: i18n.t('Day') },
    { value: WEEK, label: i18n.t('Week') },
    { value: MONTH, label: i18n.t('Month') },
    { value: YEAR, label: i18n.t('Year') },
]

export default intervals
