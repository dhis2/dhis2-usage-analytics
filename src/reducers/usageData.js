import { LOADING, ERROR } from '../constants/statuses'
import {
    APP_LOAD_SUCCESS,
    USAGE_DATA_REQUESTED,
    USAGE_DATA_RECEIVED,
    USAGE_DATA_ERRORED,
} from '../actions/types'

export default function usageData(state = LOADING, { type, payload }) {
    switch (type) {
        case USAGE_DATA_REQUESTED:
            return LOADING
        case APP_LOAD_SUCCESS:
            return payload
        case USAGE_DATA_RECEIVED:
            return payload
        case USAGE_DATA_ERRORED:
            return ERROR
        default:
            return state
    }
}
