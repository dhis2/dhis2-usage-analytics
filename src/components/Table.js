import i18n from '@dhis2/d2-i18n'
import PropTypes from '@dhis2/prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { CircularLoader } from '@dhis2/ui'
import { FAVORITE_VIEWS } from '../constants/categories'
import { TABLE_FIELDS as FIELDS } from '../constants/fields'
import {
    date as dateField,
    created as createdField,
} from '../constants/properties'
import { ERROR, LOADING } from '../constants/statuses'
import Error from './Error'
import {
    getDisplayDateForInterval,
    getDisplayDateFromIsoString,
} from '../utils/date'
import './Table.css'

const baseClassName = 'uua-data-container'
const loadingClassName = ' uua-data-container--loading'

export const TableHead = ({ headers }) => (
    <thead>
        <tr>
            {headers.map((header, i) => (
                <th key={`h-${i}`}>{header}</th>
            ))}
        </tr>
    </thead>
)

export const TableBody = ({ rows }) => (
    <tbody>
        {rows.map((row, i) => (
            <TableRow key={`row-${i}`} cells={row} />
        ))}
    </tbody>
)

export const TableRow = ({ cells }) => (
    <tr>
        {cells.map((text, i) => (
            <td key={`cell-${i}`}>{text}</td>
        ))}
    </tr>
)

export function Table({ error, loading, tableData }) {
    let content

    if (error) {
        content = (
            <Error
                message={i18n.t(
                    'There was an error retrieving the usage analytics data'
                )}
            />
        )
    } else if (loading) {
        content = <CircularLoader overlay />
    } else {
        const { headers, rows } = tableData
        content = (
            <table className="uaa-data-table">
                <TableHead headers={headers} />
                <TableBody rows={rows} />
            </table>
        )
    }

    const className = loading ? baseClassName + loadingClassName : baseClassName

    return <div className={className}>{content}</div>
}

Table.propTypes = {
    error: PropTypes.bool.isRequired,
    loading: PropTypes.bool.isRequired,
    tableData: PropTypes.object,
}

TableHead.propTypes = {
    headers: PropTypes.arrayOf(PropTypes.string).isRequired,
}

TableBody.propTypes = {
    rows: PropTypes.arrayOf(PropTypes.array).isRequired,
}

TableRow.propTypes = {
    cells: PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ).isRequired,
}

export function mapStateToProps({ usageData, filter }) {
    const error = usageData === ERROR
    const loading = usageData === LOADING
    return {
        error,
        loading,
        tableData: loading || error ? null : parseTableData(filter, usageData),
    }
}

function parseTableData({ aggregationLevel, category, interval }, dataPoints) {
    const fields =
        category === FAVORITE_VIEWS
            ? FIELDS[category][aggregationLevel]
            : FIELDS[category]

    return {
        headers: fields.map(({ label }) => label),
        rows: dataPoints.map(dataPoint =>
            fields.map(field => {
                if (field.key === dateField.key) {
                    return getDisplayDateForInterval(dataPoint, interval)
                }
                if (field.key === createdField.key) {
                    return getDisplayDateFromIsoString(dataPoint[field.key])
                }
                return dataPoint[field.key]
            })
        ),
    }
}

export default connect(mapStateToProps)(Table)
