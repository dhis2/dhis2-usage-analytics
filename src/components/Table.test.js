import React from 'react'
import { shallow } from 'enzyme'
import { Table, TableHead, TableBody, TableRow, mapStateToProps } from './Table'
import { LOADING, ERROR } from '../constants/statuses'
import * as CATS from '../constants/categories'
import * as AGGRS from '../constants/aggregations'
import { WEEK } from '../constants/intervals'
import { dataStatistics, topFavorites } from '../__mockData__/usageData'

const defaultState = {
    filter: {
        aggregationLevel: AGGRS.SUM,
        category: CATS.FAVORITE_VIEWS,
        interval: WEEK,
    },
    usageData: dataStatistics,
}

const getWrapperForState = ({ filter, usageData }) => {
    const state = {
        filter: { ...defaultState.filter, ...filter },
        usageData: usageData || defaultState.usageData,
    }
    const props = mapStateToProps(state)
    return shallow(<Table {...props} />)
}

describe('<Table/>', () => {
    it('Renders a <CircularLoader/> when usageData equals LOADING', () => {
        const wrapper = getWrapperForState({ usageData: LOADING })
        expect(wrapper.find('CircularLoader').length).toEqual(1)
    })
    it('Renders an <Error/> when usageData equals ERROR', () => {
        const wrapper = getWrapperForState({ usageData: ERROR })
        expect(wrapper.find('Error').length).toEqual(1)
    })
    it('Produces the correct table for category FAVORITE_VIEWS and aggregationLevel SUM', () => {
        const wrapper = getWrapperForState({})
        expect(wrapper).toMatchSnapshot()
    })
    it('Produces the correct table for category FAVORITE_VIEWS and aggregationLevel AVERAGE', () => {
        const wrapper = getWrapperForState({
            filter: { aggregationLevel: AGGRS.AVERAGE },
        })
        expect(wrapper).toMatchSnapshot()
    })
    it('Produces the correct table for category FAVORITES_SAVED', () => {
        const wrapper = getWrapperForState({
            filter: { category: CATS.FAVORITES_SAVED },
        })
        expect(wrapper).toMatchSnapshot()
    })
    it('Produces the correct table for category USERS', () => {
        const wrapper = getWrapperForState({ filter: { category: CATS.USERS } })
        expect(wrapper).toMatchSnapshot()
    })
    it('Produces the correct table for category TOP_FAVORITES', () => {
        const wrapper = getWrapperForState({
            filter: { category: CATS.TOP_FAVORITES },
            usageData: topFavorites,
        })
        expect(wrapper).toMatchSnapshot()
    })
    it('Produces the correct table for category DATA_VALUES', () => {
        const wrapper = getWrapperForState({
            filter: { category: CATS.DATA_VALUES },
        })
        expect(wrapper).toMatchSnapshot()
    })
})

const tableData = {
    headers: ['Header 1', 'Header 2'],
    rows: [['value a1', 'value a2'], ['value b1', 'value b2']],
}

describe('<TableHead />', () => {
    it('Matches the snapshot', () => {
        const wrapper = shallow(<TableHead headers={tableData.headers} />)
        expect(wrapper).toMatchSnapshot()
    })
})

describe('<TableBody />', () => {
    it('Matches the snapshot', () => {
        const wrapper = shallow(<TableBody rows={tableData.rows} />)
        expect(wrapper).toMatchSnapshot()
    })
})

describe('<TableRow />', () => {
    it('Matches the snapshot', () => {
        const wrapper = shallow(<TableRow cells={tableData.rows[0]} />)
        expect(wrapper).toMatchSnapshot()
    })
})
