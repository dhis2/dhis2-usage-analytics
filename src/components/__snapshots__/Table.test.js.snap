// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table/> Produces the correct table for category DATA_VALUES 1`] = `
<div
  className="uua-data-container"
>
  <table
    className="uaa-data-table"
  >
    <TableHead
      headers={
        Array [
          "Date",
          "Data Values",
        ]
      }
    />
    <TableBody
      rows={
        Array [
          Array [
            "Week 22 / 2016",
            11751,
          ],
          Array [
            "Week 23 / 2016",
            11520,
          ],
        ]
      }
    />
  </table>
</div>
`;

exports[`<Table/> Produces the correct table for category FAVORITE_VIEWS and aggregationLevel AVERAGE 1`] = `
<div
  className="uua-data-container"
>
  <table
    className="uaa-data-table"
  >
    <TableHead
      headers={
        Array [
          "Date",
          "Average Map",
          "Average Chart",
          "Average Pivot Table",
          "Average Event Report",
          "Average Event Chart",
          "Average Dashboard",
          "Average",
        ]
      }
    />
    <TableBody
      rows={
        Array [
          Array [
            "Week 22 / 2016",
            19,
            22,
            18,
            20,
            20,
            28,
            104,
          ],
          Array [
            "Week 23 / 2016",
            17,
            22,
            18,
            20,
            20,
            26,
            101,
          ],
        ]
      }
    />
  </table>
</div>
`;

exports[`<Table/> Produces the correct table for category FAVORITE_VIEWS and aggregationLevel SUM 1`] = `
<div
  className="uua-data-container"
>
  <table
    className="uaa-data-table"
  >
    <TableHead
      headers={
        Array [
          "Date",
          "Map",
          "Chart",
          "Pivot Table",
          "Event Report",
          "Event Chart",
          "Dashboard",
          "Data Set Report",
          "Total",
        ]
      }
    />
    <TableBody
      rows={
        Array [
          Array [
            "Week 22 / 2016",
            4457,
            5047,
            4281,
            4649,
            4716,
            6460,
            4414,
            23669,
          ],
          Array [
            "Week 23 / 2016",
            4093,
            5179,
            4275,
            4856,
            4782,
            6193,
            4242,
            23840,
          ],
        ]
      }
    />
  </table>
</div>
`;

exports[`<Table/> Produces the correct table for category FAVORITES_SAVED 1`] = `
<div
  className="uua-data-container"
>
  <table
    className="uaa-data-table"
  >
    <TableHead
      headers={
        Array [
          "Date",
          "Maps",
          "Charts",
          "Pivot Tables",
          "Event Reports",
          "Event Charts",
          "Dashboards",
          "Indicators",
        ]
      }
    />
    <TableBody
      rows={
        Array [
          Array [
            "Week 22 / 2016",
            1658,
            1723,
            1642,
            1748,
            1613,
            1873,
            1826,
          ],
          Array [
            "Week 23 / 2016",
            1593,
            1668,
            1643,
            1745,
            1672,
            1853,
            1905,
          ],
        ]
      }
    />
  </table>
</div>
`;

exports[`<Table/> Produces the correct table for category TOP_FAVORITES 1`] = `
<div
  className="uua-data-container"
>
  <table
    className="uaa-data-table"
  >
    <TableHead
      headers={
        Array [
          "Position",
          "Name",
          "Views",
          "ID",
          "Created",
        ]
      }
    />
    <TableBody
      rows={
        Array [
          Array [
            1,
            "ANC: ANC 3 coverage this year (gauge)",
            1,
            "dw62KezChat",
            "Nov 10, 2014",
          ],
          Array [
            2,
            "Immunization: IPT 1 and 2 Coverage",
            1,
            "nwqNUFG9Vky",
            "Jul 31, 2012",
          ],
        ]
      }
    />
  </table>
</div>
`;

exports[`<Table/> Produces the correct table for category USERS 1`] = `
<div
  className="uua-data-container"
>
  <table
    className="uaa-data-table"
  >
    <TableHead
      headers={
        Array [
          "Date",
          "Active Users",
          "Total Users",
        ]
      }
    />
    <TableBody
      rows={
        Array [
          Array [
            "Week 22 / 2016",
            227,
            262,
          ],
          Array [
            "Week 23 / 2016",
            234,
            262,
          ],
        ]
      }
    />
  </table>
</div>
`;

exports[`<TableBody /> Matches the snapshot 1`] = `
<tbody>
  <TableRow
    cells={
      Array [
        "value a1",
        "value a2",
      ]
    }
    key="row-0"
  />
  <TableRow
    cells={
      Array [
        "value b1",
        "value b2",
      ]
    }
    key="row-1"
  />
</tbody>
`;

exports[`<TableHead /> Matches the snapshot 1`] = `
<thead>
  <tr>
    <th
      key="h-0"
    >
      Header 1
    </th>
    <th
      key="h-1"
    >
      Header 2
    </th>
  </tr>
</thead>
`;

exports[`<TableRow /> Matches the snapshot 1`] = `
<tr>
  <td
    key="cell-0"
  >
    value a1
  </td>
  <td
    key="cell-1"
  >
    value a2
  </td>
</tr>
`;
